### ISSUE TITLE
Type guard with set.__contains__

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#671]
I want pyright not to raise error for the following example:

```python
def f(x: Optional[Literal["a", "b"]]):
    if x in {"a", "b"}:
        b: str = x  # Type "None" cannot be assigned to type "str"
```

In the example, `x` in if-block cannot be `None` because it is a member of `Set[str]`, but pyright cannot detect it.
Is it technically possible to solve this issue?

**Comment by erictraut** [comment]
Thanks for the suggestion. This type guard wouldn't be based on `__getitem__`. The "in" and "not in" operators use methods `__contains__` or `__iter__`.

**Comment by erictraut** [comment]
I looked into this more. Unfortunately, there's not enough information provided to the type checker to determine whether it's possible to narrow the type in this case.

The type of x in your example above is `Optional[Literal["a", "b"]]`. The type of `{"a", "b"}` is `Set[Literal["a", "b"]]`. The class `Set` is an alias for class `set`. The class `set` has a method `__contains__` that is defined as `def __contains__(self, o: object) -> bool: ...`. Since `None` is assignable to `object`, it's a legal value to pass to this method. Without assuming internal knowledge about the behavior of the `set` class (which is something a type checker should avoid), there's no way to determine from the types alone whether this method will return True or False.

**Comment by thisisrandy** [comment]
I'll add for posterity that the same logic applies to `list` and `tuple`, which are commonly used to the right of `in`:

```python
class list(MutableSequence[_T], Generic[_T]):
    def __contains__(self, __o: object) -> bool: ...

class tuple(Sequence[_T_co], Generic[_T_co]):
    def __contains__(self, __x: object) -> bool: ...
```

There's a closely related problem of why something like the following doesn't work:

```python
@dataclass
class Foo:
    bar: Literal["baz", "qux"]

def make_foo(bar: str) -> Foo:
    # fails as a guard
    assert bar in ["baz", "qux"]
    # works
    assert bar == "baz" or bar == "qux"
    return Foo(bar)
```

I think the answer, though please correct me if I'm wrong, is that `list.__contains__` is a black box or near to it from the type checker's perspective, so without making special cases for common datatypes, it can't reason about what `in` is going to return, however obvious the answer might be to the programmer.
